### ISSUE TITLE
[Enhancement] Allow groups to support multiple="false"

### ISSUE THREAD

**Issue by christian4423** [geomoose/gm3#351]
![Imgur](https://i.imgur.com/0KPrIU7.png)

I have modified **3.2.2** and I needed the catalog to support multiple=false. 

I was going to create a pull request with the change but it appears the structure for the catalog has changed in newer versions. 

Anyways, below I detail how I added this enhancement in 3.2.2

I changed the renderTreeNode() to check the multiple attr of the parent layer. 

catalog.jsx
```
renderTreeNode(childId) {
    var node = this.props.catalog[childId];
    var parent = null;
    if (node.parent !== undefined && node.parent !== null && typeof node.parent === "string" && node.parent.trim() !== "") {
        parent = this.props.catalog[node.parent];
    }
    if (node.children) {
        return this.renderGroup(node);
    } else {

        if (this.searchable && this.shouldLayerRender(node)) {
            if (parent !== undefined && parent !== null) {
                return this.renderLayer(node, parent.multiple);
            } else {
                return this.renderLayer(node);
            }
        } else {
            return false;
        }
    }
}
```

I added a whole new method in catalog.jsx that either returns a checkbox or radio. I passed it everything that was being used to render the input and the multiple boolean value getting sent to renderLayer from renderTreeNode. This new method gets called from renderLayer()
```
renderInput(multiple, doNothing, toggle, is_on) {
    if(multiple){
        return (    <input
            className="checkbox"
            type="checkbox"
            onChange={doNothing}
            onClick={toggle}
            checked={is_on}
        />)
    }else{
        return <input
        className="checkbox"
        type="radio"
        onChange={doNothing}
        onClick={toggle}
        checked={is_on}
    />
    }        
}
```

renderLayer has a few changes added to it. 

1. new param added. Defaults to true
2. toggle() basically manages the click event when multiple = false. (1) If false and other layer is clicked, it will set the other layers in the group to off and set the clicked layer to on. (2) the layer that is selected is the current layer that is on so do nothing. (3) keep default behavior as if it were a checkbox.
3. the if condition just above the return checks the is_on and multiple properties to make sure everything renders correctly. 
4. in the render method, a call to this.renderInput() is made. 

```
renderLayer(layer, multiple = true) {
    let toggle = () => {
        const map_sources = this.props.store.getState().mapSources;
        let is_on = isLayerOn(map_sources, layer);
        if(!multiple && !is_on){
            const parent = this.props.catalog[layer.parent];
            let layers = [];
            for(let child of parent.children){
                if(child !== layer.id){
                    layers.push(this.props.catalog[child])
                }
            }
            // turn off other layers
            for(let _layer of layers){
                if(isLayerOn(this.props.mapSources, _layer)){
                    this.renderMapSources(_layer, false);
                }                
            }
            // turn this on on
            this.renderMapSources(layer, true);
        }else if (!multiple && is_on){
            // do nothing
        }else{
            // default
            this.renderMapSources(layer, !isLayerOn(map_sources, layer));
        }
    };

    let toggleFavorite = () => {
        this.toggleFavoriteLayer(layer);
    };

    let toggleRefresh = () => {
        this.toggleRefreshLayer(layer);
    }

    // this prevents a React warning about the
    //  checkboxes not having an onChange event.
    let doNothing = () => { };

    let layer_classes = ['layer'];

    let is_favorite = 'favorite icon';
    if (!this.isFavoriteLayer(layer)) {
        is_favorite += ' not';
    }
    if (layer.on) {
        layer_classes.push('on');
    }

    // TODO: Check layer for minscale/maxscale against
    //       the mapView.
    // if(isOutOfScale(layer)) {
    //  layer_classes.push('out-of-scale');
    // }

    // only show the refresh icon when a layer has been configured
    //  with the ability to do auto-refresh.
    let refresh_tool = '';
    if (layer.refresh !== null && layer.refresh > 0) {
        const refresh_on = layer.refreshEnabled ? 'on' : '';
        refresh_tool = (<i className={'refresh icon ' + refresh_on} onClick={toggleRefresh} />);
    }

    let tools = this.getToolsForLayer(layer);

    let legend = false;
    if (layer.legend) {
        legend = (<Legend store={this.props.store} layer={layer} />);
    } else {
        legend = (<div className="catalog-legend"></div>)
    }

    let metadata_tool = false;
    if (layer.metadata_url) {
        metadata_tool = <MetadataTool href={layer.metadata_url} />
    }
    // check to see if the layer is on or not.
    let is_on = isLayerOn(this.props.mapSources, layer);
    const layerid = layer.id;
    if(!multiple && is_on){
        const parent = this.props.catalog[layer.parent];
        let layers = [];
        for(let child of parent.children){
            if(child !== layerid){
                layers.push(this.props.catalog[child])
            }
        }

        for(let layer of layers){
            if(isLayerOn(this.props.mapSources, layer)){
                is_on = false;
            }                
        }
    }
    return (
        <div key={layer.id} className={layer_classes.join(' ')}>
            <div className="layer-label">
                {this.renderInput(multiple, doNothing, toggle, is_on) }
                <i className={ is_favorite } onClick={toggleFavorite}/>
                <span onClick={toggle}>
                    {layer.label}
                </span>
                {refresh_tool}
                {metadata_tool}
            </div>
            <div className="layer-tools">
                <div className="tools">
                    {tools}
                </div>
                <div className="legend-spacer">
                    {legend}
                </div>
            </div>
        </div>
    );
}
```

**Comment by klassenjs** [comment]
It looks like it is working for you and you aren't the first to have wanted this in 3.x.

However, we didn't bring this forward from 2.x because it tended to cause a lot of problems if other features were used at the same time (one map-source->layer controlled by multiple catalog->layers, or if a catalog->layer controlled multiple map-source->layers).  It also interacted poorly with the visible/active layers tab. This led to a number of impossible to fix bugs in the code (whatever we did to fix one case broke it for someone else) and user confusion due to the site administrator being able to allow the user to select "impossible" conditions.  We discovered that this set of features are fundamentally in conflict (at a design level) and this couldn't be remedied with implementation changes, so we dropped it from 3.x.

My guess is "multiple=false" may be safe if it is only used when there is a one-to-one relationship between the catalog->layers and map-source->layers (and possibly for some specific cases where they aren't one-to-one.)
